### PR TITLE
fix route hooks typings

### LINKED ISSUE #4376: Wrong typings for hooks on route options
The in the snippet below, the hook types are a union of two identical types like `onRequestHookHandler<...> | onRequestHookHandler<...>` etc, and this excludes the async hook type, which should be valid to use here, right?

Are these supposed to be <code>onRequestHookHandler<...> | onRequest<em><b>Async</b></em>HookHandler<...></code>, etc?

https://github.com/fastify/fastify/blob/5c75f4993255a7e3fc5afbdc4b8171be52ecb618/types/route.d.ts#L48-L56

### PR DESCRIPTION
**Pull request by aradwann**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

fixes #4376 and  #4595

### CODE REVIEW

**Review by mcollina — APPROVED** [review]

**Review by mcollina — CHANGES_REQUESTED** [review]
The types are still incorrect.

The hook system support either the use of promises or callbacks, but not both at the same time:

`async (req, reply) => {}`

or

`(req, reply. done) => {}

File: test/types/hooks.test-d.ts [hunk 1]
```diff
@@ -256,40 +256,82 @@ type CustomContextConfig = FastifyContextConfig & {
...
+    expectType<CustomContextConfig>(request.context.config)
+    expectType<CustomContextConfig>(reply.context.config)
+  }
+})
+
+server.route<RouteGenericInterface, CustomContextConfig>({
+  method: 'GET',
+  url: '/',
+  handler: () => { },
+  onRequest: async (request, reply, done) => {
```
Comment (mcollina) [inline test/types/hooks.test-d.ts:302]: This is incorrect. When `async` is specified there should be no `done` callback.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/types/hooks.test-d.ts (hunk 1 above)
Comment (AliakbarETH) [reply test/types/hooks.test-d.ts:302]: Jees, why is the guy adding done or error payload in the code even when they are not being used in the function body?

If possible, can I correct this one?

Comment (mcollina) [reply test/types/hooks.test-d.ts:302]: Sure thing, send a fresh PR.

**Pull request closed without merging**